File: hu/codes/go/chapter_computational_complexity/worst_best_time_complexity.go
// File: worst_best_time_complexity.go
// Created Time: 2022-12-13
// Author: msk397 (machangxinq@gmail.com), cathay (cathaycchen@gmail.com)

package chapter_computational_complexity

import (
	"math/rand"
)

/* { 1, 2, ..., n } elemekből álló tömb generálása véletlenszerű sorrendben */
func randomNumbers(n int) []int {
	nums := make([]int, n)
	// nums = { 1, 2, 3, ..., n } tömb generálása
	for i := 0; i < n; i++ {
		nums[i] = i + 1
	}
	// Tömbelemek véletlenszerű összekeverése
	rand.Shuffle(len(nums), func(i, j int) {
		nums[i], nums[j] = nums[j], nums[i]
	})
	return nums
}

/* Az 1-es szám indexének keresése a nums tömbben */
func findOne(nums []int) int {
	for i := 0; i < len(nums); i++ {
		// Ha az 1-es elem a tömb elején van, a legjobb időbonyolultság O(1) érhető el
		// Ha az 1-es elem a tömb végén van, a legrosszabb időbonyolultság O(n) érhető el
		if nums[i] == 1 {
			return i
		}
	}
	return -1
}

File: hu/codes/c/chapter_computational_complexity/worst_best_time_complexity.c
/**
 * File: worst_best_time_complexity.c
 * Created Time: 2023-01-03
 * Author: codingonion (coderonion@gmail.com)
 */

#include "../utils/common.h"

/* { 1, 2, ..., n } elemeket tartalmazó, véletlenszerűen kevert tömb generálása */
int *randomNumbers(int n) {
    // Halommemória foglalása (1D változó hosszú tömb létrehozása: n db int típusú elem)
    int *nums = (int *)malloc(n * sizeof(int));
    // nums = { 1, 2, 3, ..., n } tömb generálása
    for (int i = 0; i < n; i++) {
        nums[i] = i + 1;
    }
    // Tömbelemek véletlenszerű megkeverése
    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    return nums;
}

/* Az 1-es szám indexének keresése a nums tömbben */
int findOne(int *nums, int n) {
    for (int i = 0; i < n; i++) {
        // Ha az 1-es elem a tömb elején van, a legjobb időbonyolultság O(1) teljesül
        // Ha az 1-es elem a tömb végén van, a legrosszabb időbonyolultság O(n) teljesül
        if (nums[i] == 1)
            return i;
    }
    return -1;
}

/* Vezérlő kód */
int main(int argc, char *argv[]) {
    // Véletlenszám generátor inicializálása
    srand((unsigned int)time(NULL));
    for (int i = 0; i < 10; i++) {
        int n = 100;
        int *nums = randomNumbers(n);
        int index = findOne(nums, n);
        printf("\nArray [ 1, 2, ..., n ] after shuffling = ");
        printArray(nums, n);
        printf("Index of number 1 is %d\n", index);
        // Halommemória felszabadítása
        if (nums != NULL) {
            free(nums);
            nums = NULL;
        }
    }

    return 0;
}
